src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,910 @@
+        // server challenge from Challenge msg
+        private byte[] serverChallenge = new byte[NTLM_SERVER_CHALLENGE_LENGTH];
+
+        /**
+         * Section 3.1.5.1.2 Client Receives a CHALLENGE_MESSAGE from the Server
+         * 
+         * Save messages for calculating MIC
+         */
+        private byte[] negotiateMsg = null;
+        private byte[] challengeMsg = null;
+
+        /**
+         * Creates an NTLM client context
+         *
+         * @param con
+         *        connection to SQL server
+         * @param domainName
+         *        domain name to authentication in using NTLM
+         * @param userName
+         *        user name
+         * @param password
+         *        password
+         * @param workstation
+         *        hostname of the workstation
+         * @throws SQLServerException
+         *         if error occurs
+         */
+        NTLMContext(final SQLServerConnection con, final String domainName, final String userName,
+                final String password, final String workstation) throws SQLServerException {
+
+            this.domainName = null != domainName ? domainName.toUpperCase()
+                                                 : SQLServerDriverStringProperty.DOMAIN.getDefaultValue();
+            this.domainUbytes = unicode(this.domainName);
+
+            this.userNameUbytes = null != userName ? unicode(userName) : null;
+            this.upperUserName = userName.toUpperCase();
+
+            this.passwordHash = null != password ? md4(unicode(password)) : null;
+
+            this.workstation = workstation;
+
+            String spn = null != con ? getSpn(con) : null;
+            this.spnUbytes = null != spn ? unicode(spn) : null;
+
+            try {
+                mac = Mac.getInstance("HmacMD5");
+            } catch (NoSuchAlgorithmException e) {
+                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmHmacMD5Error"));
+                Object[] msgArgs = {domainName, e.getMessage()};
+                throw new SQLServerException(form.format(msgArgs), e);
+            }
+        }
+    };
+
+    // Handle to NTLM context
+    private NTLMContext context = null;
+
+    /**
+     * Creates an instance of the NTLM authentication
+     *
+     * @param con
+     *        connection to SQL server
+     * @param domainName
+     *        domain name to authentication in using NTLM
+     * @param userName
+     *        user name
+     * @param password
+     *        password
+     * @param workstation
+     *        hostname of the workstation
+     * @throws SQLServerException
+     *         if error occurs
+     */
+    NTLMAuthentication(final SQLServerConnection con, final String domainName, final String userName,
+            final String password, final String workstation) throws SQLServerException {
+        if (null == context) {
+            this.context = new NTLMContext(con, domainName, userName, password, workstation);

```

Comment:
writing it this way (differentiating between context and this.context) makes me think that `context` was passed to this method, but it's not...it's not a big deal though.